pypose/module/ukf.py

```diff
@@ -0,0 +1,319 @@
+
+        # compute root of P
+        repeat_dim = tuple(torch.cat([torch.tensor([self.dim * 2]), torch.tensor(P.shape)]).numpy())  # repeat dim
+        p_expand = P.expand(repeat_dim)
+        np_expand = self.dim * p_expand  # calculate np
+        np_expand_sqrt = msqrt(np_expand)  # square root of np
+        np_expand_sqrt[self.dim:] *= -1
+        np_select = np_expand_sqrt.gather(1, index_gather)  # select point from np_expand_sqrt
+
+        # compute sigma point
+        x_sigma = x + np_select
+        y_sigma = bmv(C, x_sigma) + bmv(D, u) + c2
+
+        return x_sigma.reshape(-1, self.dim), y_sigma.reshape(-1, self.dim)
+
+    def compute_conv_mix(self, x_estimate, x_sigma, y_estimate, y_sigma):
+        r'''
+        compute mix covariance
+
+        Args:
+            x_estimate (:obj:`Tensor`):  System prior state estimation.
+            x_sigma (:obj:`Tensor`):  Sigma point of estimation.
+            y_estimate (:obj:`Tensor`): Observation  estimation.
+            y_sigma (:obj:`Tensor`): Observation estimation of sigma point .
+
+        Return:
+            out (:obj:`Tensor`): Mix covariance pf prior state estimation   .
+        '''
+
+        e_x = torch.sub(x_sigma, x_estimate).unsqueeze(2)
+        e_y = torch.sub(y_sigma, y_estimate).unsqueeze(2)
+        p_estimate = torch.sum(torch.bmm(e_x, e_y.permute(0, 2, 1)), dim=0)
+
+        return self.weight * p_estimate
+
+    def compute_conv(self, estimate, sigma, noise=0):
+        r'''
+        compute covariance
+
+        Args:
+            estimate (:obj:`Tensor`): System estimate.
+            sigma (:obj:`Tensor`):  Sigma point of estimate.
+
+        Return:
+            out (:obj:`Tensor`): the covariance of  prior state estimation .
+        '''
+
+        e = torch.sub(sigma, estimate).unsqueeze(2)
+        p_estimate = torch.sum(torch.bmm(e, e.permute(0, 2, 1)), dim=0)
+
+        return self.weight * p_estimate + noise
+
+    def forward(self, x, y, u, P, Q=None, R=None):
+        r'''
+        Performs one step estimation.
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            y (:obj:`Tensor`): system observation at current step (measurement)
+            u (:obj:`Tensor`): system input at current step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+            Q (:obj:`Tensor`, optional): covariance of system transition model
+            R (:obj:`Tensor`, optional): covariance of system observation model
+
+        Return:
+            list of :obj:`Tensor`: posteriori state and covariance estimation
+        '''
+        # Upper cases are matrices, lower cases are vectors
+        self.model.set_refpoint(state=x, input=u)
+        A, B = self.model.A, self.model.B
+        C, D = self.model.C, self.model.D
+        c1, c2 = self.model.c1, self.model.c2
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
```

Comment:
we don't use the linearized system for UKF. please remove.

Reply:
Should we keep $Q$ and $R$ to add noise for covariance ? $$\mathbf{P}_{y} = \frac{1}{2n} \sum\_{i=1}^{2n} \left ( \mathbf{y}^{(i)} - \hat{\mathbf{y}} \right ) \left ( \mathbf{y}^{(i)} -\hat{\mathbf{y} } \right ) ^T+\mathbf{R}$$